text/3503-frontmatter.md

```diff
@@ -0,0 +1,488 @@
+
+Downsides
+- From talking to a teacher, users are more forgiving of not understanding the details for structure data in an unstructured format (doc comments / comments) but something that looks meaningful, they will want to understand it all requiring dealing with all of the concepts
+ - The attribute approach requires explaining multiple "advanced" topics: One teacher doesn't get to teaching any attributes until the second level in his crash course series and two teachers have found it difficult to teach people raw strings
+- Attributes look "scary" (and they are in some respects for the hidden stuff they do)
+- Either we expose `syn`s lesser parse errors or we skip errors, deferring to rustc's, but then have the wrong behavior on commands that don't invoke rustc, like `cargo metadata`
+
+
+### Alternative 4: Presentation Streams
+
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+}
+
+---Cargo.toml
+[package]
+edition = "2018"
+```
+YAML allows several documents to be concatenated together variant
+[presentation streams](https://yaml.org/spec/1.2.2/#323-presentation-stream)
+which might seem familiar as this is frequently used in static-site generators
+for adding frontmatter to pages.
+What if we extended Rust's syntax to allow something similar?
+
+Benefits
+- Flexible for other content
+
+Downsides
+- Difficult to parse without assistance from something like `syn` as we'd need to distinguish what the start of a stream is vs content of a string literal
+- Being new syntax, there would be a lot of details to work out, including
+  - How to delineate and label documents
+  - How to allow escaping to avoid conflicts with content in a documents
+  - Potentially an API for accessing the document from within Rust
+- Unfamiliar, new syntax, unclear how it will work out for newer users
+
+### Alternative 5: Regular Comment
+
+Simple header:
+```rust
+#!/usr/bin/env cargo
+/* Cargo.toml:
+[package]
+edition = "2018"
+*/
+
+fn main() {
+}
+```
+
+HEREDOC:
+```rust
+#!/usr/bin/env cargo
+/* Cargo.TOML >>>
+[package]
+edition = "2018"
+<<<
+*/
+
+fn main() {
+}
+```
+The manifest can be a regular comment with a header.  If we were to support
+multiple types of content (manifest, lockfile), we could either use multiple
+comments or HEREDOC.
+This does not prescribe the exact syntax used or supported comments
+
+Benefits
+
+Downsides
+- Unfamiliar syntax
+- New style of structured comment for the ecosystem to support with potential
+  compatibility issues, likely requiring a new edition
+- Assuming it can't be parsed with `syn` and either we need to write a
+  sufficiently compatible comment parser or pull in a much larger rust parser
+  to extract and update comments.
+  - Like with doc comments, this should map to an attribute and then we'd just start the MVP with that attribute
+
+### Alternative 6: Static-site generator frontmatter
+
+```rust
+#!/usr/bin/env cargo
+---
+[package]
+edition = "2018"
+---
+
+fn main() {
+}
+```
```

Comment:
@Aloso (moving here for easier following of the thread)

> I feel like 'front matter' would be a better syntax:
> 
>     * People know that it has to be the first thing in the file, which is even implied from the name. With `` ```cargo ``, it is not immediately obvious that it can't appear in the middle or at the end of a file
> 
>     * It better conveys the purpose of the syntax: Front matter is used to add metadata to a file, while markdown code blocks are typically used to highlight some content as code
> 
>     * The proposed syntax only allows a subset of CommonMark code blocks. The CommonMark spec allows both `` ``` `` and `~~~` as code fences; it allows code fences to be indented by up to 3 spaces; and it allows the closing code fence to be longer than the opening code fence. Deviating from the CommonMark spec could lead to confusion, and implementing the spec would add a lot of complexity. Front matter on the other hand is much easier to parse correctly (if the content is just parsed as "raw text" by rustc)
> 
>     * The proposed syntax blurs the line between source code and documentation comments (where markdown is already supported). So people might incorrectly assume that doc-tests can appear outside of doc comments, or that `` ```cargo `` also works inside a doc comment.
> 
> 
> To address the mentioned downsides:
> 
> > * Too general that people might abuse it
> 
> I actually like that it is more general (it doesn't use the word 'cargo'), so it can be used for other build systems as well. Rust _the language_ shouldn't require any particular build system. What is considered 'abuse' is subjective; allowing front matter might enable various other use cases, and I wouldn't call them 'wrong' just because they're not what this feature was originally intended for.
> 
> > * We've extended the frontmatter syntax, undoing some of the "familiarity" benefit
> 
> On the other hand, familiarity with markdown syntax might backfire because the _semantics_ are so different here.
> 
> > * People are used to YAML going in frontmatter (though some systems allow other syntaxes)
> 
> We could use `+++` front matter instead of `---`.
> 
> > * Doesn't feel very rust-like
> 
> Care to elaborate?

Reply:
> The proposed syntax only allows a subset of CommonMark code blocks.

I was going off of memory of the spec.  0e671bf9 updates it to align with the spec or call out where we intentionally diverge.

> I actually like that it is more general (it doesn't use the word 'cargo'), so it can be used for other build systems as well. Rust the language shouldn't require any particular build system. What is considered 'abuse' is subjective; allowing front matter might enable various other use cases, and I wouldn't call them 'wrong' just because they're not what this feature was originally intended for.

Note that we called out adding an infostring to it and through working on this option, we'd like switch it where the infostring is required and we have the option of removing it in the future.  So this benefit is negated.

Note also that we extended fenced code-block like rules to this which also deviates from standard what people expect for YAML frontmatter

> > Doesn't feel very rust-like
>
> Care to elaborate?

A T-lang person I spoke to felt it didn't stylistically fit / match / align with Rust syntax or the Rust workflow.  I actually agree and this (along with our extensions of frontmatter syntax) are the main reasons we switched from this being our primary solution to fenced code blocks.

Reply:
> A T-lang person I spoke to felt it didn't stylistically fit / match / align with Rust syntax or the Rust workflow

I understand, but if you mention it as a downside, you should be able to explain why. Design decisions shouldn't be based on gut feelings.

Reply:
While I can't speak for that person (hoping they'll pipe in soon), for myself, I agree (though I had been quieting my different concerns over just wanting "something").  It doesn't fit in with the style of other syntax used.  This is a subjective measure though at times we have to rely on subjective measures.